### PR TITLE
Annotate Extensions.Options to make it linker friendly

### LINKED ISSUE #40236: The options subsystem needs linker attributes to preserve constructors on options types
### Description

Looking into what it means to takes to make ASP.NET Core more linkable once we get this (https://github.com/dotnet/runtime/pull/40227) and (https://github.com/dotnet/aspnetcore/pull/24458), options is the next big hurdle to cross. The issue is the very similar to DI in that the option system creates instances but the linker doesn't understand that and removes constructors from the poco objects.

```C#
public class MyOptions
{
    public int Value { get; set; };
}
```

```C#
var services = new ServiceCollection();
services.AddOptions();
services.Configure<MyOptions>(o =>
{
     o.Value = 100;
});
var service= services.GetService<IOptions<MyOptions>>();
var value = service.Value;
```

Linking this should preserve `MyOptions` constructors.

This matters for:
- `IOptions<T>`
- `IOptionsMonitor<T>`
- `IOptionsSnapshot<T>`

cc @eerhardt

### PR DESCRIPTION
**Pull request by eerhardt**
I choose to annotate the following interfaces and their implementations:

* IOptions
* IOptionsFactory
* IOptionsMonitorCache
* IOptionsMonitor
* IOptionsSnapshot

I didn't annotate `AddOptions<TOptions>`, `OptionsBuilder<TOptions>`, `Configure`, `PostConfigure`, or `Validate` because these operations don't actually create an Options class - they just consume one. It is only once someone tries getting an Options from the ServiceProvider when the class actually gets instantiated. And they way they get the Options is through one of the above annotated interfaces.

Fix #40236

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Options/tests/TrimmingTests/ConfigureTests.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        services.AddOptions<OptionsB>()
+            .Configure<IOptions<OptionsA>>((b, a) =>
+            {
+                b.OptionString = a.Value.OptionValue.ToString();
+            });
+
+        ServiceProvider provider = services.BuildServiceProvider();
+
+        OptionsA optionsA = provider.GetService<IOptions<OptionsA>>().Value;
+        OptionsB optionsB = provider.GetService<IOptions<OptionsB>>().Value;
```
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.Options/tests/TrimmingTests/ConfigureTests.cs:27]: IOptionsMonitor<OptionsB>

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Options/tests/TrimmingTests/ConfigureTests.cs]: fixed.

**Review by davidfowl — APPROVED** [review]
SWEET 😄

**Comment by eerhardt** [comment]
Failure is #40115.

**Review by maryamariyan — APPROVED** [review]
I think CI also shows apicompat error for linker tests

e.g.: 
```
.packages/microsoft.dotnet.apicompat/5.0.0-beta.20381.6/build/Microsoft.DotNet.ApiCompat.targets(82,5): error : (NETCORE_ENGINEERING_TELEMETRY=Build) TypesMustExist : Type 'Microsoft.Extensions.DependencyInjection.DefaultServiceProviderFactory' does not exist in the implementation but it does exist in the contract. [/Users/runner/work/1/s/src/libraries/Microsoft.Extensions.DependencyInjection/src/Microsoft.Extensions.DependencyInjection.csproj]
```

**Comment by eerhardt** [comment]
> I think CI also shows apicompat error for linker tests

re-running. I don't think my change caused these errors.

**Pull request merged**